### PR TITLE
Switch to pyproject.toml for build metadata

### LINKED ISSUE #154: Move to `pyproject.toml`?
Having all of the build config in `pyproject.toml` is now well supported, and now even editable installs work without `setup.py` since `pip 22`. PEP 621 defines some standard keys, so people can use new and weird build tools like poetry or flit if they like.

I'm happy to implement this if it would be useful!

### PR DESCRIPTION
**Pull request by ZedThree**
- [x] Closes #154 
- [x] Replaces #151  
- [ ] Updated `doc/whats-new.md` with a summary of the changes

This doesn't completely remove `setup.py` due to the use of `versioneer`. Switching to [`setuptools_scm`](https://github.com/pypa/setuptools_scm) would solve this, but it looks like it would a couple of other changes where `get_version` is used (`mesh.py` and the version test).

### CODE REVIEW

**Comment by johnomotani** [comment]
I'd be happy to switch to `setuptools_scm` at the same time as doing this. @ZedThree do you think it's a simple update?

**Comment by johnomotani** [comment]
@ZedThree sounds good, I don't expect performance of `get_version()` would be an issue for us!

**Comment by ZedThree** [comment]
Ok, this was a little trickier than I first thought because there's some functionality that `versioneer` provides that `setuptools_scm` doesn't -- the big one is `full-revisionid`. If the version isn't exactly on a tag, then the short (7-digit) hash will be present in the version, but if it _is_ on a tag, then the version is just the tag.

We can just run git to get the full hash of course, but then this will only work locally from a git repo.

Personally, I think only have the hash for non-tagged versions is a minor annoyance but fine. @johnomotani you might feel differently and I'm happy to defer to you if you want to just keep versioneer!

I think readthedocs is failing because the vendored versioneer needs updating to be able to read the config from pyproject.toml

**Comment by johnomotani** [comment]
@ZedThree I agree with you, I think the tag+short hash is just a minor annoyance, and not worth stopping this upgrade for.

A possible workarounds, if you have the time to try it: https://github.com/pypa/setuptools_scm says there is a `git_describe_command` option, which might make the hash format tweakable? Someone has an example here https://github.com/pandas-dev/pandas/issues/39415#issuecomment-1189354996 (in that thread there is some comment about a 'template' as well - I'm not sure but that might be a different issue to do with `pandas` wanting some specific format in `_version.py`).

**Comment by johnomotani** [comment]
I've switched hypnotoad to `setuptools_scm` now. The workaround linked in my previous comment works to get the full git hash. Slightly annoyingly, I can't find a way to get the 'dirty' status, which `versioneer` does give. Always checking the `git diff`, seems to work (including handling the case when `git diff` returns an error because the package was (non-editable) installed, so there is no git repo).

**Review by johnomotani — APPROVED** [review]

**Comment by johnomotani** [comment]
I'm going to remove the change to `setuptools_scm`. The big downside is: for an editable install (`pip install -e .`), `setuptools_scm` seems to give the version, and therefore the git hash, at the time when hypnotoad was *installed*, which is no good for an editable install. `versioneer` seems to be able to handle dynamically setting the version, so it picks up the *current* state of the repo at all times.

**Comment by johnomotani** [comment]
Needed to update the `.readthedocs.yml` to avoid some very bug-like behaviour. RTD apparently defaults to using `setup.py install` to install the package at the moment: https://github.com/readthedocs/readthedocs.org/issues/7030.

File: doc/.readthedocs.yml [hunk 1]
```diff
@@ -10,4 +10,7 @@ sphinx:
 
 python:
   install:
+    # Install doc-building requirements (sphinx, etc.)
     - requirements: docs/requirements.txt
+    # Install hypnotoad
+    - path: package
```
Comment (ZedThree) [inline doc/.readthedocs.yml:16]: I _think_ this needs to be

```suggestion
    - path: .
    - method: pip
```

Comment (johnomotani) [reply doc/.readthedocs.yml]: It's being very annoying. One thread said that the default if you include `-path:` in the install section is to use pip. That didn't seem to work, which was why I tried 'package'. Current guess is that putting two things in the same `install:` section didn't work (although I think that's what the example in the RTD docs said...).

**Comment by johnomotani** [comment]
Does RTD cache some config, or just use it from `master`?? I've just told it to not even bother installing from `doc/requirements.txt`, but it's still doing that, and `setup.py install`-ing hypnotoad. I'm tempted to just merge with the best guess at rtd config, and see if having it on `master` fixes things...

**Comment by ZedThree** [comment]
I _thought_ it used the config on the branch. Here's the config I use on ford:

```yaml
# Required
version: 2

build:
  os: ubuntu-20.04
  tools:
    python: "3.8"

sphinx:
   configuration: docs/conf.py

python:
  install:
    - method: pip
      path: .[docs]
  system_packages: true
```

**Comment by ZedThree** [comment]
Also, I think I have a working setup for using `setuptools_scm` with editable installs that does The Right Thing, but I suspect that sticking with versioneer is probably the easiest thing to do, given that it definitely works!

**Pull request merged**